tests/README.md

```diff
@@ -0,0 +1,8 @@
+﻿# Tests
```

Comment:
As we're writing the 'example' here, we need to decide if we want tests to be ran against the compiled product, or against each individual file.

I'm personally for each test being directly ran against it's associated file, so that tests can be done atomically instead of having to compile and run against the entire package. There's some things that make that impossible (classes, I'm looking at you), but it's a style choice that we need to be aware of. If you have to recompile the package and then target the compiled source instead of just running `Invoke-Pester` against the tests directory, we need to explain _why_.

Your current example module is the second style, where we have to build stuff and dot-source it using another script to actually run the tests vs using invoke-pester directly. Is this how we want to progress this as a team?

Reply:
Granted, you have classes in this module, which is a big reason to do it the way you're doing it.

Reply:
Don't we potentially have the same issue with anything we need to mock private functions for, or with references to other functions? Or do we just assume / work towards importing / dotsourcing all relevant functions for a given test?

Reply:
The way I've done it in the two places where I've done it that way is that the test imports anything and everything it needs to finish the test. So public functions would import private files if needed. This is where things get a bit messy, but it also helps with bookkeeping because the test itself does the bookkeeping of what is tied to what.

Reply:
### The _best practice_ is the one that works _all the time_.  

It might be _convenient_ to run tests by dot-sourcing the function you're testing, but it's not the best practice, because frequently it cannot be done, or cannot be done in ways that _also_ allow testing the assembled module.

* When you depend on classes
* When you depend on module-scope variables or configuration
* When testing private functions (which requires `InModuleScope`)

In particular, the test itself should not dot-source the SUT (Subject Under Test), because that ties the test to the layout of the source code _and prevents using it to test the assembled module_.

However, even when none of these problems are in evidence, and you can easily write tests against an individual function -- teaching people to do it that way can lead to tests that _only_ work that way, and cannot be used to test the assembled module, just because the developers never run them that way themselves. For instance, developers may simply write _implementations_ of internal dependencies, instead of using mocks -- because the function they need to mock isn't defined unless they also dot-source that!

I'm OK with individuals manually running specific tests against individual functions, but writing tests that work _both ways_ is an extra level of complication -- and not something we need to teach in the best practices.

Remember the point of this is to teach people to build modules for others to use. **All the tests** need to run against the complete module after it's been put together. Code coverage and test correctness can _only_ be evaluated in that context, so the tests _must_ all work properly that way.

In the CI build we really should be able to build the module, copy the tests to the output folder, and on a clean system, import the module and run all the tests -- and get 95% code coverage 😉

Reply:
Joel's argument makes far more sense than mine. Do it his way.